File: day11/Python/day11.py
from termcolor import colored
from pyfiglet import *
print(colored("Advent of Code - Day 11", "yellow").center(80, "-"))
print(colored(figlet_format("Dumbo Octopus",font="small",justify="center"), 'green'))
print(colored("Output","yellow").center(80, "-"))

f = open("../Input/day11.txt","r")
list_1 = []
for line in f:
    list_1.append([int(x) for x in line.strip()])


R = len(list_1)
C = len(list_1[0])

ans = 0
def flash(r,c):
    global ans
    ans += 1
    list_1[r][c] = -1
    for dr in [-1,0,1]:
        for dc in [-1,0,1]:
            rr = r+dr
            cc = c+dc
            if 0<=rr<R and 0<=cc<C and list_1[rr][cc]!=-1:
                list_1[rr][cc] += 1
                if list_1[rr][cc] >= 10:
                    flash(rr,cc)

t = 0
while True:
    t += 1
    for r in range(R):
        for c in range(C):
            list_1[r][c] += 1
    for r in range(R):
        for c in range(C):
            if list_1[r][c] == 10:
                flash(r,c)
    done = True
    for r in range(R):
        for c in range(C):
            if list_1[r][c] == -1:
                list_1[r][c] = 0
            else:
                done = False
    if t == 100:
        print("\nPuzzle 1: ",ans)
    if done:
        print("Puzzle 2: ",t,"\n")
        break
print(colored("=".center(71, "="), "yellow"))

File: day11/Kotlin/day11.kt
import java.io.*
fun print_day_11() {
    val yellow = "\u001B[33m"
    val reset = "\u001b[0m"
    val green = "\u001B[32m"
    println(yellow + "-".repeat(25) + "Advent of Code - Day 11" + "-".repeat(25) + reset)
    println(green)
    val process = Runtime.getRuntime().exec("figlet Dumbo Octopus -c -f small")
    val reader = BufferedReader(InputStreamReader(process.inputStream))
    reader.forEachLine { println(it) }
    println(reset)
    println(yellow + "-".repeat(33) + "Output" + "-".repeat(33) + reset + "\n")
}
fun main() {
    print_day_11()
    Day11().part_1()
    Day11().part_2()
    println("\n" + "\u001B[33m" + "=".repeat(72) + "\u001b[0m" + "\n")
}

class Day11 {
    private val octogrid = File("../Input/day11.txt").readLines().map { it.toCharArray().map { c -> c.digitToInt() }.toMutableList() }.toMutableList()
    data class GridPoint(val x:Int, val y:Int) {
        fun neighbours(xBoundary: Int, yBoundary: Int): List<GridPoint> =
            (-1 .. 1).map { yOffset ->  (-1 .. 1).map { xOffset -> GridPoint(x+xOffset, y+yOffset)  } }.flatten()
                .filter {
                    it.x in 0 until xBoundary && it.y  in 0 until yBoundary && !(it.x == x && it.y == y)
                }
    }
    fun part_1() {
        var totalFlashes = 0
        for(step in 1..100) {
            val flashed = mutableListOf<GridPoint>()
            for (y in octogrid.indices) {
                for (x in octogrid[y].indices) {
                    octogrid[y][x]++
                }
            }
            do {
                val newFlashes = mutableListOf<GridPoint>()
                for (y in octogrid.indices) {
                    for (x in octogrid[y].indices) {
                        val gp = GridPoint(x, y)
                        if (octogrid[y][x] > 9 && gp !in flashed) {
                            newFlashes.add(gp)
                            gp.neighbours(octogrid[y].size, octogrid.size).forEach {
                                octogrid[it.y][it.x]++
                            }
                        }
                    }
                }
                flashed.addAll(newFlashes)
            } while(newFlashes.isNotEmpty())
            totalFlashes += flashed.size
            flashed.forEach { octogrid[it.y][it.x] = 0 }
        }
        println("Puzzle 1: " + totalFlashes)
    }
    fun part_2() {
        val totalOctos = octogrid.flatten().size
        var step = 0
        do {
            step++
            val flashed = mutableListOf<GridPoint>()
            for (y in octogrid.indices) {
                for (x in octogrid[y].indices) {
                    octogrid[y][x]++
                }
            }
            do {
                val newFlashes = mutableListOf<GridPoint>()
                for (y in octogrid.indices) {
                    for (x in octogrid[y].indices) {
                        val gp = GridPoint(x, y)
                        if (octogrid[y][x] > 9 && gp !in flashed) {
                            newFlashes.add(gp)
                            gp.neighbours(octogrid[y].size, octogrid.size).forEach {
                                octogrid[it.y][it.x]++
                            }
                        }
                    }
                }
                flashed.addAll(newFlashes)
            } while(newFlashes.isNotEmpty())
            flashed.forEach { octogrid[it.y][it.x] = 0 }
        } while(flashed.size != totalOctos)

        println("Puzzle 2: " + step)
    }
}